nim/tests/test_routes.py

```diff
@@ -0,0 +1,110 @@
+import json
+from nim.app import app
+import pytest
+
+
+@pytest.fixture
+def client_app():
+    return app.test_client()
+
+
+def assert_new_game_success(passed_client, req):
+    """
+    Asserts post to /new is 200
+    & state is not empty
+    :param passed_client: test_client
+    :param req: json to post
+    :return: bool
+    """
+    response = passed_client.post(
+        '/new',
+        content_type='application/json',
+        data=req
+    )
+    assert response.status_code == 200 and not len(json.loads(response.data)['state']) == 0
+
+
+def assert_new_game_failed(passed_client, req):
+    """
+    Asserts post tp /new gets error code 400
+     & the string 'error' exists in the response
+    :param passed_client: test_client
+    :param req: json to post
+    :return: bool
+    """
+    response = passed_client.post(
+        '/new',
+        content_type='application/json',
+        data=req
+    )
+    assert response.status_code == 400 and b'error' in response.data
+
+
+def test_index(client_app):
+    response = client_app.get('/', content_type='html/text')
+    assert response.status_code == 200
+
+
+def test_new_success(client_app):
+    assert_new_game_success(client_app,
+                            json.dumps({'min': 5, 'max': 15, 'piles': 4}))
+
+
+def test_new_omit_min_success(client_app):
+    assert_new_game_success(client_app,
+                            json.dumps({'max': 15, 'piles': 4}))
+
+
+def test_new_omit_max_success(client_app):
+    assert_new_game_success(client_app,
+                            json.dumps({'min': 5, 'piles': 4}))
+
+
+def test_new_omit_piles_success(client_app):
+    assert_new_game_success(client_app,
+                            json.dumps({'min': 5, 'max': 15}))
+
+
+def test_new_omit_all_success(client_app):
+    assert_new_game_success(client_app,
+                            json.dumps({}))
+
+
+def test_new_max_limit_fail(client_app):
+    assert_new_game_failed(client_app,
+                           json.dumps({'min': 10, 'max': 51, 'piles': 3}))
+
+
+def test_new_piles_limit_fail(client_app):
+    assert_new_game_failed(client_app,
+                           json.dumps({'min': 10, 'max': 10, 'piles': 51}))
+
+
+def test_new_all_limit_fail(client_app):
```

Comment:
I'm not convinced that `test_new_all_limit_fail` is needed.
What scenario are we protecting ourselves against that isn't already covered by `test_new_max_limit_fail` and `test_new_piles_limit_fail`? That somehow having one of those arguments be 0 would fail but having all of them at 0 would succeed?
I think you can drop this test, or maybe replace it with a test that passes a negative number to make sure we're handling that correctly too.

Reply:
Fair comment. 
The case I had in mind was testing the limit for `min`. Because if you test the limit on `min` without also putting `max` past the limit, you'll get a different error. (`min can't be greater than max').
So I wanted to make sure min reacted as expected to going above limit

Reply:
So at the minimum you should pass a valid number for `piles` then, otherwise all you can say is that passing an invalid value for `piles` fails.

Reply:
That's fair